### PR TITLE
feat(models): add GPT-5.6 Sol / Terra / Luna support

### PR DESCRIPTION
**Pull request by ndycode**
## Summary

- Adds the three GPT-5.6 tiers (`gpt-5.6-sol`, `gpt-5.6-terra`, `gpt-5.6-luna`) as first-class models, with bare `gpt-5.6` aliased to Sol.
- Adds two reasoning tiers above `xhigh`: `max` (all three tiers) and `ultra` (Sol/Terra only).
- Fixes two latent bugs found while wiring this up (details below).

Facts were taken from the upstream Codex catalog (`openai/codex` → `codex-rs/models-manager/models.json`) and `codex-rs/core/src/client.rs`, not the prose API docs — the docs pages contradict each other on whether a `max` tier exists, and quote a context window for the API surface that does not match the Codex backend.

## What Changed

- **`lib/request/helpers/model-map.ts`** — `ModelReasoningEffort` gains `max` + `ultra`; new `WireReasoningEffort` excludes `ultra`. Adds the three model profiles, per-tier effort aliases, and a dedicated 5.6 resolver.
- **`lib/request/request-transformer.ts`** — exhaustive fallback rows for `max`/`ultra`; rewrites `ultra` → `max` before the request is built; guarded effort-suffix pattern.
- **`lib/types.ts`** — `ConfigOptions.reasoningEffort` accepts `ultra`; `ReasoningConfig.effort` is `WireReasoningEffort`, so `ultra` cannot reach the wire by construction.
- **`lib/usage/pricing.ts`** — Sol $5/$30, Terra $2.50/$15, Luna $1/$6 per 1M tokens (cached input at 10% of input, matching existing entries).
- **`config/codex-modern.json`** — the three tiers, 372k context, no `none` variant, `ultra` only on Sol/Terra.
- **`docs/configuration.md`, `config/README.md`** — document the new tiers and the `ultra` → `max` behavior.
- **`test/gpt56-models.test.ts`** — new, 20 tests.

### Catalog facts encoded

| Tier | Efforts | Default |
|---|---|---|
| `gpt-5.6-sol` | low, medium, high, xhigh, max, **ultra** | `low` |
| `gpt-5.6-terra` | low, medium, high, xhigh, max, **ultra** | `medium` |
| `gpt-5.6-luna` | low, medium, high, xhigh, max | `medium` |

No tier accepts `none` or `minimal`; those coerce up rather than being sent.

### `ultra` never reaches the wire

Upstream `reasoning_effort_for_request` rewrites `Ultra → Max` **unconditionally** before sending:

```rust
fn reasoning_effort_for_request(effort: ReasoningEffortConfig) -> ReasoningEffortConfig {
    match effort {
        ReasoningEffortConfig::Ultra => ReasoningEffortConfig::Max,
        effort => effort,
    }
}
```

`ultra` denotes client-side subagent delegation, not a distinct backend effort. Sending it would be rejected. It stays selectable in config and is emitted as `max`.

### Two latent bugs fixed

1. **`gpt-5.1-codex-max` already ends in `-max`.** Naively adding `max` to the effort-suffix pattern strips it to `gpt-5.1-codex`, silently rerouting every Codex Max request. The lookbehind is scoped to the `max` alternative only (`gpt-5-codex-low` must still parse `low`). Covered by a regression test that was confirmed to fail when the guard is removed.
2. **Unrecognised `gpt-5.6-*` ids resolved to `gpt-5.5`.** The general GPT-5 resolver has no minor `6`, so it fell through to the "stable" variant — meaning `gpt-5.6-sol` silently ran 5.5 before this change. A dedicated 5.6 resolver now catches these.

## Validation

- [x] `npm run lint`
- [x] `npm run typecheck`
- [x] `npm test`
- [x] `npm test -- test/documentation.test.ts` (26 passed)
- [x] `npm run build`

Full suite: **5045 passed**, 6 skipped. The 3 failures (`ci-workflows`, `plugin-manifest`, `runtime-rotation-proxy`) are **pre-existing** — verified by stashing this branch and re-running them against clean `main` (`82b36af`), where they fail identically. None touch models, efforts, or pricing.

## Docs and Governance Checklist

- [x] README updated (if user-visible behavior changed) — n/a; model list lives in `docs/configuration.md` + `config/README.md`, both updated
- [ ] `docs/getting-started.md` updated (onboarding flow unchanged)
- [ ] `docs/features.md` updated (no new capability surface)
- [x] relevant `docs/reference/*` pages updated — `docs/configuration.md`
- [ ] `docs/upgrade.md` updated (no migration behavior change)
- [x] `SECURITY.md` and `CONTRIBUTING.md` reviewed for alignment

## Risk and Rollback

- **Risk level: low.** Purely additive for existing users. Legacy `gpt-5` still resolves to `gpt-5.5`; `DEFAULT_MODEL` and `QUOTA_PROBE_MODEL_CHAIN` are untouched, so 5.6 is strictly opt-in. The only behavior change to an existing path is the `codex-max` suffix guard, which fixes a bug that would otherwise have been introduced here.
- **Rollback:** revert the single commit; no migrations, no persisted state, no config rewrites.

## Additional Notes

Deliberately **not** included, as each changes behavior for existing users and deserves its own call:

- `DEFAULT_MODEL` still `gpt-5.5` (upstream Codex now defaults to Sol).
- `QUOTA_PROBE_MODEL_CHAIN` unchanged — live/quota probes still hit 5.5 first.
- No unsupported-model fallback chain for 5.6 → 5.5. If an account lacks 5.6 access the request fails rather than degrading. Worth adding if 5.6 access is uneven across accounts.
- No `CHANGELOG.md` entry yet.

5.6 is routed to the `gpt-5.2` prompt family because no `gpt_5_6_prompt.md` exists upstream — 5.6 ships its base instructions inline in `models.json`, and `gpt-5.5` already reuses that family the same way.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_017a48odxu9pEVoYpkw3m26V

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

adds first-class support for gpt-5.6-sol, gpt-5.6-terra, and gpt-5.6-luna with two new reasoning tiers (`max` and `ultra`), where `ultra` is a client-side alias that rewrites to `max` before the wire. also fixes two latent bugs: a codex-max suffix-stripping collision and a silent gpt-5.6-* → 5.5 fallback.

- `ModelReasoningEffort`/`WireReasoningEffort` moved to `lib/constants.ts` as a leaf module to break a latent import cycle; `VARIANT_SUFFIX_PATTERN` in `request-transformer.ts` is now derived from `REASONING_EFFORTS` so a new tier can't be silently missed
- three model profiles added with per-tier effort lists, pricing, and a dedicated `resolveGpt56CatalogModel` resolver; the lookbehind guard `(?<!codex)-(max)` prevents stripping `-max` from `gpt-5.1-codex-max`
- `lib/config.ts` `getUnsupportedCodexFallbackChain` normalizer retains a hardcoded `-(none|minimal|low|medium|high|xhigh)` pattern that was not updated — fallback chain keys written with `-max` or `-ultra` suffixes will not normalize correctly

<h3>Confidence Score: 4/5</h3>

safe to merge for all existing users; the one real defect is confined to an opt-in advanced config feature that no existing users have configured for GPT-5.6 yet

the fallback-chain normalizer in lib/config.ts carries a duplicate hardcoded effort-suffix list that wasn't updated — users who write gpt-5.6-sol-max or gpt-5.6-sol-ultra as keys in unsupportedCodexFallbackChain will silently get no-op fallback behavior because the key stays un-normalized in the map. the rest of the change is correct and well-tested

lib/config.ts line 1244 — the getUnsupportedCodexFallbackChain normalizer

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lib/constants.ts | adds REASONING_EFFORTS array and ModelReasoningEffort/WireReasoningEffort types as a leaf module; correctly excludes ultra from WireReasoningEffort |
| lib/request/helpers/model-map.ts | adds three GPT-5.6 profiles, addEffortAliases helper, addGpt56Aliases, and resolveGpt56CatalogModel; lookbehind guard on -max is correct; resolver order (codex → 5.6 → general) is safe |
| lib/request/request-transformer.ts | VARIANT_SUFFIX_PATTERN derives from REASONING_EFFORTS and guards -max with lookbehind; ultra→max rewrite before the wire is correct; REASONING_FALLBACKS extended for max/ultra |
| lib/types.ts | ConfigOptions.reasoningEffort widened to ModelReasoningEffort; ReasoningConfig.effort tightened to WireReasoningEffort (excludes ultra); module boundary respected — no import from lib/request/** |
| lib/config.ts | getUnsupportedCodexFallbackChain normalizer has a hardcoded effort-suffix regex that was not updated — max/ultra suffixes on fallback-chain keys won't be stripped, silently breaking the feature for GPT-5.6 entries |
| lib/usage/pricing.ts | adds Sol/Terra/Luna pricing at $5/$30, $2.50/$15, $1/$6 per 1M tokens; cached rates at 10% of input; reasoning rate equals output rate, matching existing pattern |
| config/codex-modern.json | adds all three 5.6 tiers at 372k/128k context/output; Sol and Terra include ultra variant; Luna correctly omits it; no none/minimal variants |
| test/gpt56-models.test.ts | 20 tests covering resolution, effort coercion, suffix parsing, profiles, and pricing; regression test for codex-max suffix guard; missing vitest coverage for config.ts fallback-chain normalizer with max/ultra |
| test/module-boundaries.test.ts | new import-cycle guard using readFileSync + regex; importSpecifiers only matches double-quoted imports so single-quoted would be invisible; safe given ESLint quote enforcement but fragile by construction |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["user config: model = 'gpt-5.6-sol-max'"] --> B["getModelConfig()\nVARIANT_SUFFIX_PATTERN strips -max\n(?<!codex)-(max)"]
    B --> C["model key = 'gpt-5.6-sol'\neffort = 'max'"]
    C --> D["normalizeModel()\nresolveNormalizedModel()"]
    D --> E{In MODEL_MAP?}
    E -- yes --> F["'gpt-5.6-sol' (exact alias)"]
    E -- no --> G["resolveGpt56CatalogModel()\ntokens includes 'terra'? luna? default→sol"]
    F --> H["getModelProfile()\nreturns GPT_5_6_SOL profile\nsupportedEfforts: low/med/high/xhigh/max/ultra"]
    G --> H
    H --> I["coerceReasoningEffort()\n'max' in supportedEfforts → 'max'"]
    I --> J{"coercedEffort === 'ultra'?"}
    J -- yes --> K["wireEffort = 'max'"]
    J -- no --> L["wireEffort = coercedEffort"]
    K --> M["ReasoningConfig { effort: 'max' }"]
    L --> M
    style B fill:#e8f5e9
    style I fill:#e8f5e9
    style K fill:#fff9c4
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["user config: model = 'gpt-5.6-sol-max'"] --> B["getModelConfig()\nVARIANT_SUFFIX_PATTERN strips -max\n(?<!codex)-(max)"]
    B --> C["model key = 'gpt-5.6-sol'\neffort = 'max'"]
    C --> D["normalizeModel()\nresolveNormalizedModel()"]
    D --> E{In MODEL_MAP?}
    E -- yes --> F["'gpt-5.6-sol' (exact alias)"]
    E -- no --> G["resolveGpt56CatalogModel()\ntokens includes 'terra'? luna? default→sol"]
    F --> H["getModelProfile()\nreturns GPT_5_6_SOL profile\nsupportedEfforts: low/med/high/xhigh/max/ultra"]
    G --> H
    H --> I["coerceReasoningEffort()\n'max' in supportedEfforts → 'max'"]
    I --> J{"coercedEffort === 'ultra'?"}
    J -- yes --> K["wireEffort = 'max'"]
    J -- no --> L["wireEffort = coercedEffort"]
    K --> M["ReasoningConfig { effort: 'max' }"]
    L --> M
    style B fill:#e8f5e9
    style I fill:#e8f5e9
    style K fill:#fff9c4
```

</a>

<sub>Reviews (3): Last reviewed commit: ["test(models): derive the suffix-pattern ..."](https://github.com/ndycode/codex-multi-auth/commit/6cffc5764d33ca8445ca50e50db54ec6555e0a28) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=43156961)</sub>

### CODE REVIEW

File: lib/request/request-transformer.ts [hunk 1]
```diff
@@ -63,6 +64,28 @@ export function normalizeModel(model: string | undefined): string {
...
+const REASONING_EFFORTS = [
+	"none",
+	"minimal",
+	"low",
+	"medium",
+	"high",
+	"xhigh",
+	"max",
+	"ultra",
+] as const satisfies readonly ModelReasoningEffort[];
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/request/request-transformer.ts]: Fixed in `cf96840`. `VARIANT_SUFFIX_PATTERN` is now built from `REASONING_EFFORTS` at module load, with `max` kept as a separate alternative for its lookbehind:

```ts
const VARIANT_SUFFIX_PATTERN = new RegExp(
  `(?:-(${REASONING_EFFORTS.filter((effort) => effort !== "max").join("|")})|(?<!codex)-(max))$`,
  "i",
);
```

Verified the generated `.source` is byte-identical to the literal it replaces, and that both produce the same capture for `gpt-5.1-codex-max`, `codex-max`, `gpt-5-codex-low`, `gpt-5.1-codex-max-xhigh`, and each 5.6 suffix.

File: lib/types.ts [hunk 2]
```diff
@@ -1,4 +1,8 @@
 import type { Auth, Provider, Model } from "@codex-ai/sdk";
+import type {
+	ModelReasoningEffort,
+	WireReasoningEffort,
+} from "./request/helpers/model-map.js";
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/types.ts]: Fixed in `cf96840`, though not quite as suggested — defining them in `lib/types.ts` and importing back from model-map would have closed an import cycle:

`lib/types.ts` → `lib/schemas.ts` → `lib/request/helpers/model-map.ts`

(`lib/schemas.ts:8` imports `MODEL_FAMILIES`/`ModelFamily` from model-map.) So model-map importing `lib/types.ts` completes the loop, which `import-x/no-cycle` is set to `error` on.

Took the parenthetical instead — the types now live in the leaf `lib/constants.ts` (no intra-repo imports), and model-map re-exports them so every existing caller is unchanged.

Also added `test/module-boundaries.test.ts` for the missing regression coverage. Worth noting `import-x/no-cycle` alone would *not* catch a regression here now: since model-map no longer imports the types layer, re-pointing `lib/types.ts` back at model-map creates no cycle. The test asserts the boundary directly, and I confirmed it fails when the import is reverted.
